Функциональность для пользователей и вещей

goloviu opened the pull request:
feat: добавить типы данных
feat: добавить функционал для работы с пользователями
feat: добавить функционал для работы с вещами

goloviu committed:
feat: добавить типы данных
feat: добавить функционал для работы с пользователями
feat: добавить функционал для работы с вещами

goloviu committed:
feat: добавить типы данных
feat: добавить функционал для работы с пользователями
feat: добавить функционал для работы с вещами

goloviu committed:
style: исправить конфликты по код стайл

goloviu committed:
fix: изменить код билдера в мапперах

irek-valiev commented on src/main/java/ru/practicum/shareit/item/dto/ItemMapper.java:
```diff
@@ -0,0 +1,34 @@
+public class ItemMapper {
+
+    public static ItemDto toItemDto(final Item item) {
+        @Valid
```
валидация здесь избыточна, сервис следит за консистентностью своих данных, а значит не даст ответ "кривыми" данными

goloviu replied:
Исправлено
Имеется ввиду, что с БД мы не можем получить некачественные данные т.к валидируем их на входе или программа постоянно следит за валидацией?

irek-valiev replied:
Именно так. Сервис перед записью в БД валидирует данные, поэтому при обратном процессе уже можно отдавать как есть

irek-valiev commented on src/main/java/ru/practicum/shareit/item/dto/ItemMapper.java:
```diff
@@ -0,0 +1,34 @@
+        return itemDto;
+    }
+
+    public static Item toItem(final ItemDto itemDto) {
```
этот метод нужно использовать при мапинге в таких методах как создание и обновление

goloviu replied:
Исправлено

irek-valiev commented on src/main/java/ru/practicum/shareit/item/dto/ItemMapper.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    public static Item toItem(final ItemDto itemDto) {
+        @Valid
```
а здесь уже была проведена эта валидация в контроллере, вторая не нужна

goloviu replied:
Исправлено

irek-valiev commented on src/main/java/ru/practicum/shareit/item/model/Item.java:
```diff
@@ -1,7 +1,27 @@
+@Builder
+@AllArgsConstructor
+@NoArgsConstructor
 public class Item {
```
валидация данных проводится в классах DTO, а в model или entity описывается модель БД, поэтому model и не афишируют. В этом классе валидацию полей нужно убрать. Когда подключится в проект JPA, на поля будут добавлены другие аннотации

goloviu replied:
Исправлено

irek-valiev commented on src/main/java/ru/practicum/shareit/item/ItemController.java:
```diff
@@ -3,10 +3,63 @@
+
+    @PostMapping
+    public ItemDto addNewItem(@RequestHeader("X-Sharer-User-Id") Long userId,
+                              @RequestBody @Valid Item item) {
```
для обмена данными используются классы DTO (data transfer object), на этих же классах проводится валидация. Поэтому методы контроллера на вход получают DTO и возвращают тоже его

goloviu replied:
Исправлено

irek-valiev commented on src/main/java/ru/practicum/shareit/item/ItemServiceImpl.java:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    @Override
+    public ItemDto addNewItem(final Long userId, final Item item) {
```
на вход методу должен прийти ItemDto, потом мапинг через метод toItem() класса ItemMapper и уже после этого вызов itemStorage.save()

goloviu replied:
Исправлено

irek-valiev commented on src/main/java/ru/practicum/shareit/user/dto/UserMapper.java:
```diff
@@ -0,0 +1,18 @@
+
+import javax.validation.Valid;
+
+public class UserMapper {
```
нужно добавить метод маппинга DTO -> model и использовать его

goloviu replied:
Исправлено

irek-valiev commented on src/main/java/ru/practicum/shareit/user/dto/UserMapper.java:
```diff
@@ -0,0 +1,18 @@
+public class UserMapper {
+
+    public static UserDto toUserDto(final User user) {
+        @Valid
```
валидация здесь избыточна, сервис следит за консистентностью своих данных, а значит не даст ответ "кривыми" данными

goloviu replied:
Исправлено

irek-valiev commented on src/main/java/ru/practicum/shareit/user/model/User.java:
```diff
@@ -0,0 +1,22 @@
+
+@Data
+@Builder
+public class User {
```
валидация данных проводится в классах DTO, а в model или entity описывается модель БД, поэтому model и не афишируют. В этом классе валидацию полей нужно убрать. Когда подключится в проект JPA, на поля будут добавлены другие аннотации

goloviu replied:
Исправлено

irek-valiev commented on src/main/java/ru/practicum/shareit/user/InMemoryUserRepositoryImpl.java:
```diff
@@ -0,0 +1,110 @@
+@Component
+@Slf4j
+public class InMemoryUserRepositoryImpl implements UserRepository {
+    private final HashMap<Long, User> users = new HashMap<>();
```
при объявлении переменных лучше использовать интерфейсы, а не их классы реализации. Т.е. интерфейс Map, а не класс HashMap

goloviu replied:
Как-то упустил этот момент, спасибо:)

goloviu replied:
исправлено

irek-valiev commented on src/main/java/ru/practicum/shareit/user/InMemoryUserRepositoryImpl.java:
```diff
@@ -0,0 +1,110 @@
+            throw new UserAlreadyExistsException("Пользователь уже существует");
+        } else if (user.getId() != null) {
+            throw new IllegalArgumentException("У нового пользователя не должен быть указан ID");
+        }
```
перед сохранением нового пользователя нужно проверить уникальность его email

goloviu replied:
Исправлено

irek-valiev commented on src/main/java/ru/practicum/shareit/user/UserController.java:
```diff
@@ -1,12 +1,49 @@
+    }
+
+    @PostMapping
+    public UserDto saveNewUser(@RequestBody @Valid User user) {
```
для обмена данными используются классы DTO (data transfer object), на этих же классах проводится валидация. Поэтому методы контроллера на вход получают DTO и возвращают тоже его

goloviu replied:
Исправлено

goloviu committed:
fix: добавить новые мапперы DTO для Item, User
fix: исправить типы данных в контроллере Item, User
fix: Добавить проверку уникальности email при сохранении юзера
